src/commands/react-native/upload.ts

```diff
@@ -0,0 +1,319 @@
+import {RequestBuilder} from '../../helpers/interfaces'
+import {getMetricsLogger, MetricsLogger} from '../../helpers/metrics'
+import {upload, UploadStatus} from '../../helpers/upload'
+import {getRequestBuilder} from '../../helpers/utils'
+import {RN_SUPPORTED_PLATFORMS, RNPlatform, RNSourcemap} from './interfaces'
+import {
+  renderCommandInfo,
+  renderConfigurationError,
+  renderFailedUpload,
+  renderGitDataNotAttachedWarning,
+  renderRetriedUpload,
+  renderSuccessfulCommand,
+  renderUpload,
+} from './renderer'
+import {InvalidPayload, validatePayload} from './validation'
+
+export class UploadCommand extends Command {
+  public static usage = Command.Usage({
+    description: 'Upload React Native sourcemaps to Datadog.',
+    details: `
+            This command will upload React Native sourcemaps and their corresponding javascript bundle to Datadog in order to un-minify front-end stack traces received by Datadog.
+            See README for details.
+        `,
+    examples: [
+      [
+        'Upload ios sourcemaps',
+        'datadog-ci react-native upload --platform ios --service com.company.app --bundle ./main.jsbundle --sourcemap ./main.jsbundle.map --release-version 1.23.4',
+      ],
+      [
+        'Upload android sourcemaps',
+        'datadog-ci react-native upload --platform android --service com.company.app --bundle ./index.android.bundle --sourcemap ./index.android.bundle.map --release-version 1.23.4',
+      ],
+    ],
+  })
+
+  private buildVersion?: string
+  private bundle?: string
+  private cliVersion: string
+  private config = {
+    apiKey: process.env.DATADOG_API_KEY,
+    datadogSite: process.env.DATADOG_SITE || 'datadoghq.com',
+  }
+  private disableGit?: boolean
+  private dryRun = false
+  private maxConcurrency = 20
+  private platform?: RNPlatform
+  private projectPath: string = process.cwd() || ''
+  private releaseVersion?: string
+  private repositoryURL?: string
+  private service?: string
+  private sourcemap?: string
+
+  constructor() {
+    super()
+    this.cliVersion = require('../../../package.json').version
+  }
+
+  public async execute() {
+    if (!this.releaseVersion) {
+      this.context.stderr.write('Missing release version\n')
+
+      return 1
+    }
+
+    if (!this.buildVersion) {
+      this.context.stderr.write('Missing build version\n')
+
+      return 1
+    }
+
+    if (!this.service) {
+      this.context.stderr.write('Missing service\n')
+
+      return 1
+    }
+
+    if (!this.bundle) {
+      this.context.stderr.write('Missing bundle path\n')
+
+      return 1
+    }
+
+    if (!this.platform) {
+      this.context.stderr.write('Missing platform\n')
+
+      return 1
+    }
+
+    if (!RN_SUPPORTED_PLATFORMS.includes(this.platform)) {
+      this.context.stderr.write(
+        `Platform ${this.platform} is not supported.\nSupported platforms are ios and android.\n`
```

Comment:
Could we make this documentation future proof by using the list instead of manually encoding the possible values in the string?
Also even if that's outside of the scope of the PR I feel we could have a quick win by providing shared helpers functions to validate (but I don't want to super-charge this PR too much)
```typescript
try {
    validateRequiredField("--bundle-path", this.bundle)
    validateOptionField("--platform", this.platform, RN_SUPPORTED_PLATFORMS)
} catch (e) {
    this.context.stderr.write(e)

    return 1
}
```

Reply:
I agree it'd be a lot easier to understand what `execute` does like this.
In the meantime I would like to have this code merged soon to avoid conflicts if someone makes changes to the `sourcemaps` command or starts a global refactoring.

Is it ok for you if I open a new PR to do the change once this one is merged?

Reply:
yes 👍